core/sail/fts/lucene-spin/src/main/java/org/eclipse/rdf4j/lucene/spin/LuceneSpinSailConnection.java

```diff
@@ -0,0 +1,261 @@
+	 */
+	protected final SailConnectionListener connectionListener = new SailConnectionListener() {
+
+		@Override
+		public void statementAdded(Statement statement) {
+			// we only consider statements that contain literals
+			if (statement.getObject() instanceof Literal) {
+				if (statement == null)
+					return;
+				// we further only index statements where the Literal's datatype is
+				// accepted
+				Literal literal = (Literal)statement.getObject();
+				if (luceneIndex.accept(literal))
+					buffer.add(statement);
+			}
+		}
+
+		@Override
+		public void statementRemoved(Statement statement) {
+			// we only consider statements that contain literals
+			if (statement.getObject() instanceof Literal) {
+				if (statement == null)
+					return;
+				// we further only indexed statements where the Literal's datatype
+				// is accepted
+				Literal literal = (Literal)statement.getObject();
+				if (luceneIndex.accept(literal))
+					buffer.remove(statement);
+			}
+		}
+	};
+
+	/**
+	 * To remember if the iterator was already closed and only free resources once
+	 */
+	private final AtomicBoolean closed = new AtomicBoolean(false);
+
+	public LuceneSpinSailConnection(NotifyingSailConnection wrappedConnection, SearchIndex luceneIndex) {
+		super(wrappedConnection);
+		this.luceneIndex = luceneIndex;
+
+		/*
+		 * Using SailConnectionListener, see <a href="#whySailConnectionListener">above</a>
+		 */
+
+		wrappedConnection.addConnectionListener(connectionListener);
+	}
+
+	@Override
+	public synchronized void addStatement(Resource arg0, IRI arg1, Value arg2, Resource... arg3)
+		throws SailException
+	{
+		super.addStatement(arg0, arg1, arg2, arg3);
+	}
+
+	@Override
+	public void close()
+		throws SailException
+	{
+		if (closed.compareAndSet(false, true)) {
+			try {
+				super.close();
+			}
+			finally {
+				try {
+					luceneIndex.endReading();
+				}
+				catch (IOException e) {
+					logger.warn("could not close IndexReader or IndexSearcher " + e, e);
+				}
+				// remember if you were closed before, some sloppy programmers
+				// may call close() twice.
+			}
+		}
+	}
+
+	// //////////////////////////////// Methods related to indexing
+
+	@Override
+	public synchronized void clear(Resource... arg0)
+		throws SailException
+	{
+		// remove the connection listener, this is safe as the changing methods
+		// are synchronized
+		// during the clear(), no other operation can be invoked
+		getWrappedConnection().removeConnectionListener(connectionListener);
+		try {
+			super.clear(arg0);
+			buffer.clear(arg0);
+		}
+		finally {
+			getWrappedConnection().addConnectionListener(connectionListener);
+		}
+	}
+
+	@Override
+	public void begin()
+		throws SailException
+	{
+		super.begin();
+		buffer.reset();
+		try {
+			luceneIndex.begin();
+		}
+		catch (IOException e) {
+			throw new SailException(e);
+		}
+	}
+
+	@Override
+	public void commit()
+		throws SailException
```

Comment:
For both `buffer` and `SailIindex` operations are thread safe. I assume all that things are solved there. That method also was copied from another class and I believe it was tested. If not maybe it is worth to create unit test for that.